src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubDirectDStream.java

```
@@ -0,0 +1,268 @@
+    LOG.debug("Starting a batch.");
+    if (!takeSnapshot) {
+      this.currentSnapshotName = existingSnapshot;
+      return;
+    }
+
+    String subscription = config.getSubscription();
+    this.currentSnapshotName = createAndSave(config.getProject(), subscription,
+                                             subscription + "-snapshot-" + UUID.randomUUID(), context);
+  }
+
+  @Override
+  public void onBatchRetry(io.cdap.cdap.etl.api.streaming.StreamingContext streamingContext) {
+    LOG.debug("Batch is about to be retried. Seeking to snapshot {} for the current batch.", currentSnapshotName);
+    seekSnapshot(config.getProject(), config.getSubscription(), currentSnapshotName);
+  }
+
+  @Override
+  public void onBatchCompleted(io.cdap.cdap.etl.api.streaming.StreamingContext streamingContext) {
+    try {
+      streamingContext.deleteSate(config.getSubscription());
+    } catch (IOException e) {
+      throw new RuntimeException("Deleting state failed. ", e);
+    }
+    // delete the snapshot from Pub/Sub
+    deleteSnapshot(currentSnapshotName);
+
+    // For next batch
+    takeSnapshot = true;
+    LOG.debug("Batch completed successfully. Deleted snapshot {} for the current batch.", currentSnapshotName);
+  }
+
+  private void createSubscriptionIfNotPresent() {
+    try {
+      PubSubSubscriberUtil.createSubscription(() -> true, BackoffConfig.defaultInstance(),
+                                              ProjectSubscriptionName.format(config.getProject(),
+                                                                             config.getSubscription()),
+                                              TopicName.format(config.getProject(), config.getTopic()),
+                                              () -> subscriptionAdminClient,
+                                              PubSubSubscriberUtil::isApiExceptionRetryable);
+    } catch (Exception e) {
+      throw new RuntimeException("Subscription creation failed.", e);
```

think it would be better to let the exception propagate up instead of wrapping it here.

Seems like it's only called from start(), what is the behavior if start() throws an exception?

---

Wrapping as a runtime exception since this is an overridden method. Exception thrown from start() causes the pipeline to fail.